pkg/controller/vmicollision/vmicollision_controller.go

```diff
@@ -0,0 +1,369 @@
+
+// PoolManagerInterface defines the methods required by VMICollisionReconciler
+type PoolManagerInterface interface {
+	IsVirtualMachineManaged(namespace string) (bool, error)
+	GetManagedVirtualMachineNamespaces() (map[string]struct{}, error)
+}
+
+// VMICollisionReconciler watches VirtualMachineInstance objects and detects MAC address collisions
+type VMICollisionReconciler struct {
+	client.Client
+	poolManager PoolManagerInterface
+	recorder    record.EventRecorder
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func SetupWithManager(mgr manager.Manager, poolManager *pool_manager.PoolManager) error {
+	// Create the reconciler
+	r := &VMICollisionReconciler{
+		Client:      mgr.GetClient(),
+		poolManager: poolManager,
+		recorder:    mgr.GetEventRecorderFor("kubemacpool"),
+	}
+
+	// Register the MAC address indexer
+	if err := mgr.GetFieldIndexer().IndexField(
+		context.Background(),
+		&kubevirtv1.VirtualMachineInstance{},
+		MacAddressIndexName,
+		IndexVMIByMAC,
+	); err != nil {
+		return fmt.Errorf("failed to setup MAC address indexer: %w", err)
+	}
+
+	log.Info("Successfully registered MAC address indexer for VMI collision detection")
+
+	c, err := controller.New("vmicollision-controller", mgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create VMI collision controller: %w", err)
+	}
+
+	err = c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&kubevirtv1.VirtualMachineInstance{},
+			&handler.TypedEnqueueRequestForObject[*kubevirtv1.VirtualMachineInstance]{},
+			collisionRelevantChanges(),
+		),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to watch VMIs: %w", err)
+	}
+
+	log.Info("Successfully registered VMI watch with collision-relevant predicates")
+
+	startupChan := make(chan event.GenericEvent, 1000)
+
+	if err := c.Watch(
+		source.Channel(startupChan, &handler.EnqueueRequestForObject{}),
+	); err != nil {
+		return fmt.Errorf("failed to watch startup channel: %w", err)
+	}
+
+	if err := mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
+		logger := log.WithName("startup-enqueue")
+		logger.Info("Waiting for informer cache to sync")
+
+		if !mgr.GetCache().WaitForCacheSync(ctx) {
+			select {
+			case <-ctx.Done():
+				logger.Info("Context cancelled during cache sync, shutting down startup enqueuing")
+			default:
+				logger.Error(nil, "Cache sync failed, skipping startup collision detection")
+			}
+			return nil
+		}
+
+		logger.Info("Cache synced successfully, enqueuing Running VMIs for collision detection")
+		if err := enqueueAllRunningVMIs(ctx, r.Client, r.poolManager, startupChan, logger); err != nil {
+			logger.Error(err, "Failed to enqueue VMIs at startup, continuing anyway")
+		}
+		return nil
+	})); err != nil {
+		return fmt.Errorf("failed to add startup enqueuer: %w", err)
+	}
+
+	log.Info("Successfully registered startup enqueuing for VMI collision detection")
+	return nil
```

Comment:
Are we sure we really need this ? the controller runtime is supose to do this on start by itself.